src/hotspot/share/opto/loopnode.cpp

```diff
@@ -4721,102 +4810,121 @@ void PhaseIdealLoop::verify_compare( Node *n, const PhaseIdealLoop *loop_verify,
-    while( *pp != loop )
-      pp = &((*pp)->_next);
-    // Find proper sibling to be next
-    IdealLoopTree **nn = &loop->_next;
-    while( (*nn) && (*nn)->_head != _head )
-      nn = &((*nn)->_next);
-
-    // Check for no match.
-    if( !(*nn) ) {
-      // Annoyingly, irreducible loops can pick different headers
-      // after a major_progress operation, so the rest of the loop
-      // tree cannot be matched.
-      if (_irreducible && Compile::current()->major_progress())  return;
-      assert( 0, "failed to match loop tree" );
-    }
-
-    // Move (*nn) to (*pp)
-    IdealLoopTree *hit = *nn;
-    *nn = hit->_next;
-    hit->_next = loop;
-    *pp = loop;
-    loop = hit;
-    // Now try again to verify
-  }
-
-  assert( _head  == loop->_head , "mismatched loop head" );
+bool IdealLoopTree::verify_tree(IdealLoopTree* loop_verify) const {
+  assert(_head == loop_verify->_head, "mismatched loop head");
+  assert(this->_parent != nullptr || this->_next == nullptr, "is_root_loop implies has_no_sibling");
+
+  // Collect the children
+  GrowableArray<IdealLoopTree*> children;
+  GrowableArray<IdealLoopTree*> children_verify;
+  collect_children(children);
+  loop_verify->collect_children(children_verify);
+  children.sort(compare_tree);
+  children_verify.sort(compare_tree);
+
+  bool fail = false;
+
+  // Compare the two children lists
+  for (int i = 0, j = 0; i < children.length() || j < children_verify.length(); ) {
+    IdealLoopTree* child        = nullptr;
+    IdealLoopTree* child_verify = nullptr;
+    // Read from both lists, if possible.
+    if (i < children.length()) {
+      child = children.at(i);
+    }
+    if (j < children_verify.length()) {
+      child_verify = children_verify.at(j);
+    }
+    if (child != nullptr && child_verify != nullptr && child->_head != child_verify->_head) {
+      assert(child->_head->_idx != child_verify->_head->_idx, "is implied");
+      // We found two non-equal children. Select the smaller one.
+      if (child->_head->_idx < child_verify->_head->_idx) {
+        child_verify = nullptr;
+      } else {
+        child = nullptr;
+      }
+    }
+    // Process the two children, or potentially log the failure if we only found one.
+    if (child_verify == nullptr) {
+      if (child_verify->_irreducible && Compile::current()->major_progress()) {
+        // Irreducible loops can pick a different header (one of its entries).
+      } else {
+        tty->print_cr("We have loop that verify does not have");
+        child->dump();
+        fail = true;
+      }
+      i++; // step for this
+    } else if (child == nullptr) {
+      if (child_verify->_irreducible && Compile::current()->major_progress()) {
+        // Irreducible loops can pick a different header (one of its entries).
+      } else if (child_verify->_head->as_Region()->is_in_infinite_subgraph()) {
+        // Infinite loops do not get attached to the loop-tree on their first visit.
```

Comment:
Can you explain why you don't check `child` for infinite subgraph?

Reply:
As the comment here says, infinite loops are not attached the first time we run `build_loop_tree`.
But from the second time on, we have the `NeverBranch` added, so it has a "fake" exit, and is not really an infinite loop any more, so it will always be found and attached to the loop tree.
`this` runs before `loop_verify`, so it is possible that `child` does not have the loop, but `child_verify` finds it. It is not possible the other way around.
I can add a comment to the code for that.

Reply:
Side note:
Before my patch, we did not check if `loop_verify` had any loops that we did not have in `this`. That is why we did not have to deal with suddenly appearing infinite loops.

Reply:
Yes, please add comment.